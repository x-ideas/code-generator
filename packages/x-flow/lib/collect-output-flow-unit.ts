/*
 * @name:
 * description:
 *  用于收集所有子flow unit的output的数据，统一传递给下一个flow unit
 */

import { XGroupedFlowUnit } from './grouped-flow-unit';

export class CollectOutputFlowUnit extends XGroupedFlowUnit {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
  async doWork(data: any): Promise<any[]> {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const result: any[] = [];

    let nextData = data;
    for (const unit of this._units) {
      nextData = await unit.doWork(nextData);
      result.push(nextData);
    }

    return result;
  }
}
