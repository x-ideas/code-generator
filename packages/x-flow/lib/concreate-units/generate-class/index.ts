/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  将interface生成前端class定义的FlowUnit
 */

import * as babelParser from '@babel/parser';
import lodash from 'lodash';

import { XFlowUnit } from '../../flow-unit';

import jscodeshift from 'jscodeshift';
import { Collection } from 'jscodeshift/src/Collection';
import * as K from 'jscodeshift/node_modules/ast-types/gen/kinds';
import { parserConfig } from '../../utils/jscodeshift-parser';

export class GenerateClassFlowUnit extends XFlowUnit {
  #convertedFromNiceFormat: boolean;
  constructor({ convertedFromNiceFormat = true }: { convertedFromNiceFormat: boolean }) {
    super();

    this.#convertedFromNiceFormat = convertedFromNiceFormat;
  }

  /**
   *
   * @private
   * @param {Collection<jscodeshift.>} collection
   * @returns {Collection<ASTNode>}
   * @memberof GenerateClassFlowUnit
   */
  private _changeInterfaceToClass(collection: Collection<jscodeshift.ASTNode>): Collection<jscodeshift.ASTNode> {
    const newCollection = jscodeshift('');
    const newProgram = newCollection.find(jscodeshift.Program);

    collection.find(jscodeshift.TSInterfaceDeclaration).forEach(path => {
      const classDeclaration = generateClassDeclaration(path, {
        convertedFromNiceFormat: this.#convertedFromNiceFormat,
      });
      newProgram.get('body').value.push(classDeclaration);
    });

    return newCollection;
  }

  async doWork(interfaceSources: string): Promise<string> {
    const collection: Collection<jscodeshift.ASTNode> = jscodeshift(interfaceSources, {
      parser: parserConfig(),
    });
    // const a = collection.find(jscodeshift.TSInterfaceDeclaration).get('body').get('body');
    // // console.log(collection.find(jscodeshift.TSInterfaceDeclaration).length);

    // // const ds = collection.find(jscodeshift.TSInterfaceDeclaration);
    // // const id = ds.get('id');
    // console.log(collection.find(jscodeshift.TSInterfaceDeclaration).get('id').get('name').value);

    return this._changeInterfaceToClass(collection).toSource();
  }
}

/**
 * 生成exporse装饰器
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {jscodeshift.Decorator}
 */
function generateExposeDecorator(
  property: jscodeshift.TSPropertySignature,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.Decorator {
  // 找到参数
  const argums: jscodeshift.ObjectExpression[] = [];
  if (jscodeshift.Identifier.check(property.key)) {
    // property

    const bName: string = options.convertedFromNiceFormat ? lodash.snakeCase(property.key.name) : property.key.name;

    const nameProperty = jscodeshift.property('init', jscodeshift.identifier('name'), jscodeshift.literal(bName));
    argums.push(jscodeshift.objectExpression([nameProperty]));
  }

  return jscodeshift.decorator(jscodeshift.callExpression(jscodeshift.identifier('Expose'), argums));
}

/**
 * 生成validate装饰器
 * TODO: 判断是否是可选的属性
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {jscodeshift.Decorator}
 */
function generateValidateDecorator(property: jscodeshift.TSPropertySignature): jscodeshift.Decorator | null {
  // let name: string = '';
  // if (jscodeshift.Identifier.check(property.key)) {
  //   name = property.key.name;
  // }

  // if (!name) {
  //   return null;
  // }

  if (property.optional) {
    return null;
  }

  if (property.typeAnnotation && jscodeshift.TSTypeAnnotation.check(property.typeAnnotation)) {
    // TODO: 根据类型和name生成对应的call expression

    switch (property.typeAnnotation.typeAnnotation.type) {
      // number, enum当作number
      case 'TSNumberKeyword': {
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(
              jscodeshift.identifier.from({
                comments: [jscodeshift.commentBlock('NOTE: 判断为number的属性可能为枚举，如果为枚举，请手动增加一个枚举定义并替换该属性的类型')],
                name: 'getDefaultValidNum',
              }),
              []
            ),
          ])
        );

        break;
      }

      case 'TSStringKeyword': {
        // 判断是否为id // string;
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(jscodeshift.identifier('getDefaultValidStr'), []),
          ])
        );
      }

      case 'TSBooleanKeyword':
        // boolean
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('ValidateValue'), [
            jscodeshift.callExpression(jscodeshift.identifier('getDefaultValidBoolean'), []),
          ])
        );

      case 'TSTypeReference':
        // 另一个类型定义(interface, class, enum),
        // NOTE: 通过前缀区分
        return jscodeshift.decorator(
          jscodeshift.callExpression(jscodeshift.identifier('Type'), [
            // @ts-ignore
            jscodeshift.arrowFunctionExpression([], jscodeshift.identifier(property.typeAnnotation.typeAnnotation.typeName?.name)),
          ])
        );

      case 'TSArrayType':
        {
          if (property.typeAnnotation.typeAnnotation.elementType.type === 'TSTypeReference') {
            const typeName = property.typeAnnotation.typeAnnotation.elementType.typeName;
            if (jscodeshift.Identifier.check(typeName)) {
              return jscodeshift.decorator(
                jscodeshift.callExpression(jscodeshift.identifier('Type'), [
                  jscodeshift.arrowFunctionExpression([], jscodeshift.identifier(typeName.name)),
                ])
              );
            }
          }
        }

        break;

      default:
        break;
    }
  }

  return null;
}

/**
 * 生成class property的初始化值
 * @param {jscodeshift.TSPropertySignature} property
 */
function generateClassPropertyInitValue(property: jscodeshift.TSPropertySignature): K.ExpressionKind | null {
  // 找到type类型
  const optional = property.optional;
  if (optional) {
    return null;
  }

  switch (property.typeAnnotation?.typeAnnotation.type) {
    case 'TSBooleanKeyword':
      return jscodeshift.literal(false);

    case 'TSNumberKeyword':
      return jscodeshift.literal(0);

    case 'TSStringKeyword': {
      // 是否是id
      jscodeshift.Identifier.assert(property.key);
      if (jscodeshift.Identifier.check(property.key)) {
        if (property.key.name.includes('id')) {
          return jscodeshift.callExpression(jscodeshift.identifier('getFrontId'), []);
        }
      }

      return jscodeshift.stringLiteral('');
    }

    case 'TSArrayType':
      return jscodeshift.arrayExpression([]);

    case 'TSTypeReference': {
      // TODO: 根据名字去判断
      return null;
    }

    default:
      return null;
  }
}

/**
 * 生成class property的标注
 * 如果property是可选的，那么会得到一个联合类型
 * @param {jscodeshift.TSParameterProperty} property
 * @returns {TSTypeAnnotationKind}
 */
function generateClassPropertyAnnotation(property: jscodeshift.TSPropertySignature): jscodeshift.TSTypeAnnotation | undefined {
  function _generate(typeAnnotation: K.TSTypeKind | K.TSTypeAnnotationKind, optional = false): jscodeshift.TSTypeAnnotation | undefined {
    switch (typeAnnotation.type) {
      case 'TSBooleanKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsBooleanKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsBooleanKeyword());
      }
      case 'TSNumberKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsNumberKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsNumberKeyword());
      }
      case 'TSStringKeyword': {
        return optional
          ? jscodeshift.tsTypeAnnotation(jscodeshift.tsUnionType([jscodeshift.tsStringKeyword(), jscodeshift.tsUndefinedKeyword()]))
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsStringKeyword());
      }

      case 'TSArrayType': {
        return optional
          ? jscodeshift.tsTypeAnnotation(
              jscodeshift.tsUnionType([jscodeshift.tsArrayType(typeAnnotation.elementType), jscodeshift.tsUndefinedKeyword()])
            )
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsArrayType(typeAnnotation.elementType));
      }

      case 'TSTypeReference': {
        return optional
          ? jscodeshift.tsTypeAnnotation(
              jscodeshift.tsUnionType([jscodeshift.tsTypeReference(typeAnnotation.typeName), jscodeshift.tsUndefinedKeyword()])
            )
          : jscodeshift.tsTypeAnnotation(jscodeshift.tsTypeReference(typeAnnotation.typeName));
      }
    }
    return undefined;
  }

  if (property.typeAnnotation) {
    switch (property.typeAnnotation.type) {
      case 'TSTypeAnnotation': {
        return _generate(property.typeAnnotation.typeAnnotation, property.optional);
        break;
      }

      default:
        break;
    }
  }

  return undefined;
}

/**
 * 生成class的属性定义
 * @param {jscodeshift.TSPropertySignature} property
 * @returns {(jscodeshift.ClassProperty | undefined)}
 */
function generateClassPropertyFromInterfaceProperty(
  property: jscodeshift.TSPropertySignature,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.ClassProperty | null {
  if (jscodeshift.Identifier.check(property.key)) {
    const ccName = lodash.camelCase(property.key.name);

    const value = generateClassPropertyInitValue(property);
    const typeAnnotation = generateClassPropertyAnnotation(property);
    const cProperty = jscodeshift.classProperty(jscodeshift.identifier(ccName), value, typeAnnotation);
    cProperty.comments = property.comments;

    // @ts-ignore
    // NOTE: 定义居然没有
    cProperty.decorators = [generateExposeDecorator(property, options), generateValidateDecorator(property)].filter(Boolean);

    return cProperty;
  }
  return null;
}

/**
 * 生成class声明
 * @param {string} className
 * @param {jscodeshift.TSPropertySignature[]} properties
 * @returns {jscodeshift.ExportNamedDeclaration}
 */
function generateClassDeclaration(
  interfaceDeclaration: jscodeshift.ASTPath<jscodeshift.TSInterfaceDeclaration>,
  options: {
    convertedFromNiceFormat: boolean;
  }
): jscodeshift.ExportNamedDeclaration {
  const interfaceName: string = jscodeshift(interfaceDeclaration).get('id', 'name').value;
  const properties = jscodeshift(interfaceDeclaration).get('body', 'body').value;

  // @ts-ignore
  const cProperties = properties.map(prop => generateClassPropertyFromInterfaceProperty(prop, options));

  // 将第一个I替换成C
  return jscodeshift.exportNamedDeclaration(
    jscodeshift.classDeclaration(jscodeshift.identifier(`C${interfaceName.slice(1)}`), jscodeshift.classBody(cProperties))
  );
}
