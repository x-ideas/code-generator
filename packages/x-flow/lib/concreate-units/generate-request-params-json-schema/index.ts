/* eslint-disable @typescript-eslint/ban-ts-comment */
/*
 * @name:
 * description:
 *  解析swagger文档，得到request，并将其转换成json schema
 */

import { JSONSchema4 } from 'json-schema';
import { OpenAPIV2 } from 'openapi-types';
import lodash from 'lodash';

import { XFlowUnit } from '../../flow-unit';

interface IGenerateRequestParamsSchemaResult {
  path: JSONSchema4;
  body: JSONSchema4;
  query: JSONSchema4;
}

function isInBodyParameterObject(parameter: OpenAPIV2.Parameter): parameter is OpenAPIV2.InBodyParameterObject {
  return parameter.in === 'body';
}

export class GenerateRequestParamsJsonSchemaFlowUnit extends XFlowUnit {
  /**
   * 获取描述
   * @param obj
   */
  private getDescription(obj: OpenAPIV2.OperationObject) {
    if (obj.summary && obj.description) {
      return `${obj.summary} 
       ${obj.description}
      `;
    }
    return obj.summary ?? obj.description ?? '';
  }

  private _convertJsonTypeDefineFromOpenApi(parameter: OpenAPIV2.GeneralParameterObject): JSONSchema4 {
    switch (parameter.type) {
      case 'array':
        return {
          type: parameter.type,
          description: parameter.description,
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          items: this.generateItemJsonSchema(parameter.items, {}, []),
        };

      case 'object':
        return {
          type: parameter.type,
          description: parameter.description,
          properties: this.generatePropertiesJSONSchema(parameter.properties ?? {}, {}, []),
        };

      case 'string':
      case 'number':
      case 'integer':
      case 'boolean':
        return {
          type: parameter.type,
          description: parameter.description ?? '',
        };

      default:
        return {};
    }
  }

  /**
   * 解析path参数
   * @private
   * @param {OpenAPIV2.Parameter[]} parameters
   * @returns {{
   *     [name: string]: JSONSchema4;
   *   }}
   * @memberof GenerateRequestParamsJsonSchemaFlowUnit
   */
  private convertPathParemters(
    parameters: OpenAPIV2.Parameter[]
  ): {
    [name: string]: JSONSchema4;
  } {
    return (
      parameters
        .filter(par => {
          return par.in === 'path';
        })
        // @ts-ignore
        .map((par: OpenAPIV2.GeneralParameterObject) => {
          return {
            [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
          };
        })
        .reduce((accum, current) => {
          return {
            ...accum,
            ...current,
          };
        }, {})
    );
  }

  /**
   * 获得必填的path参数
   * @private
   * @param {OpenAPIV2.Parameter[]} parameters
   * @returns {string[]}
   * @memberof GenerateRequestParamsJsonSchemaFlowUnit
   */
  private getRequeiredPathParemters(parameters: OpenAPIV2.Parameter[]): string[] {
    return parameters
      .filter(par => {
        return par.in === 'path';
      })
      .filter(par => {
        return par.required === true;
      })
      .map(par => par.name);
  }

  private generateItemJsonSchema(
    items?: OpenAPIV2.ItemsObject,
    defs: OpenAPIV2.DefinitionsObject = {},
    refs: string[] = []
  ): JSONSchema4 | JSONSchema4[] | undefined {
    if (items) {
      if (items.$ref) {
        if (!refs.includes(items.$ref)) {
          // @ts-ignore
          throw Error('此时不应该出现ref，马上联系管理员');
          // this.dealWithRef(items.$ref, defs, refs, items?.description);
        }
      }

      switch (items.type) {
        case 'array': {
          return {
            type: items.type,
            items: this.generateItemJsonSchema(items.items, defs, refs),
          };
        }

        case 'object': {
          // 此时ref被填充
          return {
            type: items.type,
            // @ts-ignore
            description: items.title ?? '',
            // @ts-ignore
            properties: this.generatePropertiesJSONSchema(items.properties ?? {}, defs, refs),
          };
        }

        case 'string':
        case 'number':
        case 'integer':
        case 'boolean':
          return {
            type: items.type,
            description: 'item-暂无',
          };
      }
    }

    return undefined;
  }

  private generateJSonSchemaFromOpenAPISchema(
    schema: OpenAPIV2.SchemaObject,
    defs: OpenAPIV2.DefinitionsObject,
    refs: string[],
    description?: string
  ): JSONSchema4 {
    if (schema.$ref && !refs.includes(schema.$ref)) {
      throw Error('此时不应该存在ref, 或者这是一个错误');
    }

    switch (schema.type) {
      case 'array': {
        const items = schema.items;
        return {
          type: schema.type,
          description: schema.description ?? schema.title ?? description,
          items: this.generateItemJsonSchema(items, defs, refs),
        };
      }

      case 'object': {
        return {
          type: schema.type,
          description: schema.description ?? description,
          properties: this.generatePropertiesJSONSchema(schema.properties ?? {}, defs, refs),
        };
      }

      case 'string':
      case 'number':
      case 'integer':
      case 'boolean':
        return {
          type: schema.type,
          description: schema.description ?? description,
          // default: schema.default,
          // required: schema.required,
        };

      default:
        return {};
    }
  }

  private generatePropertiesJSONSchema(
    properties: { [key: string]: OpenAPIV2.SchemaObject },
    defs: OpenAPIV2.DefinitionsObject,
    refs: string[],
    description = ''
  ): { [key: string]: JSONSchema4 } {
    const result: { [key: string]: JSONSchema4 } = {};
    for (const key of Object.keys(properties)) {
      const property = properties[key];

      result[key] = this.generateJSonSchemaFromOpenAPISchema(property, defs, refs, description);
    }

    return result;
  }

  private convertBodyParemeters(parameters: OpenAPIV2.Parameter[]): { [name: string]: JSONSchema4 } {
    return (
      parameters
        .filter(par => {
          return par.in === 'body';
        })
        // @ts-ignore
        .map((par: OpenAPIV2.GeneralParameterObject) => {
          if (par.schema) {
            // this.generateJSonSchemaFromOpenAPISchema(par.schema, {}, []);

            // const result: JSONSchema4 = {
            //   type: par.schema.type ?? 'object',
            //   description: par.schema.description ?? par.schema.title ?? '',
            //   properties: this.generatePropertiesJSONSchema(par.schema.properties ?? {}, {}, []),
            // };
            return {
              [par.name]: this.generateJSonSchemaFromOpenAPISchema(par.schema, {}, [], par.description),
            };
          }
          return {
            [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
          };
        })
        .reduce((accum, current) => {
          return {
            ...accum,
            ...current,
          };
        }, {})
    );
  }

  private convertQueryParemeters(parameters: OpenAPIV2.Parameter[]): { [name: string]: JSONSchema4 } {
    return (
      parameters
        .filter(par => {
          return par.in === 'query';
        })
        // @ts-ignore
        .map((par: OpenAPIV2.GeneralParameterObject) => {
          return {
            [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
          };
        })
        .reduce((accum, current) => {
          return {
            ...accum,
            ...current,
          };
        }, {})
    );
  }

  private getRequeiredQueryParemters(parameters: OpenAPIV2.Parameter[]): string[] {
    return parameters
      .filter(par => {
        return par.in === 'query';
      })
      .filter(par => {
        return par.required === true;
      })
      .map(par => par.name);
  }

  /**
   * 将swagger文档中的请求参数转换成符合scheme 结构的properties
   * @private
   * @param {OpenAPIV2.Parameter[]} parameters
   * @memberof GenerateRequestJsonSchema
   */
  private convertParameters(parameters: OpenAPIV2.Parameter[]): { [k: string]: JSONSchema4 } | undefined {
    const result = parameters
      .filter(par => {
        // 过滤掉header中的参数，因为是用户权限，统一设置
        return par.in !== 'header';
      })
      // .filter(par => {
      //   // 过滤掉path中的参数，这部分参数另外处理
      //   return par.in !== 'path';
      // })
      .map((par: OpenAPIV2.Parameter) => {
        if (par.in === 'path') {
          // 路径参数
          const result: JSONSchema4 = {
            path: {
              [par.name]: this._convertJsonTypeDefineFromOpenApi(par as OpenAPIV2.GeneralParameterObject),
            },
          };
          return result;
        }

        if (isInBodyParameterObject(par)) {
          const result: JSONSchema4 = {
            body: par.schema,
          };
          return result;
        } else {
          const result: JSONSchema4 = {
            [par.in]: {
              [par.name]: this._convertJsonTypeDefineFromOpenApi(par),
            },
          };
          return result;
        }
      })
      .reduce((accum, current) => {
        return lodash.merge(accum, current);
      }, {});

    return result;
  }

  private findRequestInPath(obj: OpenAPIV2.Document): OpenAPIV2.OperationObject | undefined {
    const paths: OpenAPIV2.PathItemObject[] = Object.values(obj.paths);
    if (paths[0]) {
      const firstPath = paths[0];

      return firstPath.get ?? firstPath.post ?? firstPath.delete ?? firstPath.put ?? undefined;
    }

    return undefined;
  }

  private getRequeiredParamers(parameters: OpenAPIV2.Parameter[]): string[] {
    return parameters
      .filter(par => {
        return par.in !== 'header' && par.in !== 'path';
      })
      .filter(par => {
        return par.required === true;
      })
      .map(par => {
        if (par.in === 'body') {
          return 'body';
        } else {
          return par.name;
        }
      });
  }

  async doWork(operationObj: OpenAPIV2.Document): Promise<IGenerateRequestParamsSchemaResult> {
    // 找到path中的请求对象

    const requestDef = this.findRequestInPath(operationObj);
    if (!requestDef) {
      throw Error('没有找到请求的定义');
    }

    return {
      path: {
        $schema: 'http://json-schema.org/draft-04/schema#',
        description: this.getDescription(requestDef),
        // NOTE: 以为先用swagger parse进行处理，我们这里先不考虑ref
        properties: this.convertPathParemters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
        required: this.getRequeiredPathParemters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
      },
      body: {
        $schema: 'http://json-schema.org/draft-04/schema#',
        description: this.getDescription(requestDef),
        properties: this.convertBodyParemeters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
        required: [],
      },
      query: {
        $schema: 'http://json-schema.org/draft-04/schema#',
        description: this.getDescription(requestDef),
        properties: this.convertQueryParemeters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
        required: this.getRequeiredQueryParemters((requestDef.parameters ?? []) as OpenAPIV2.Parameter[]),
      },
    };
  }
}
