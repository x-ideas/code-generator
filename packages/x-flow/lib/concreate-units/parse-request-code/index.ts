export * from './parser-request-code-unit';
