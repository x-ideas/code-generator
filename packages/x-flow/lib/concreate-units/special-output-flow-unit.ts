import { XFlowUnit } from '../flow-unit';

/**
 * 指定输出的flow unit
 * @export
 * @class SpecialOutputFlowUnit
 * @extends {XFlowUnit}
 */
export class SpecialOutputFlowUnit extends XFlowUnit {
  private _data: any;

  constructor(data: any) {
    super();
    this.setOutputData(data);
  }

  setOutputData(data: any): void {
    this._data = data;
  }

  async doWork(): Promise<any> {
    return this._data;
  }
}
