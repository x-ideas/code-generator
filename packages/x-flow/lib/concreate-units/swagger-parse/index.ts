export * from './swagger-parser-flow-unit';
