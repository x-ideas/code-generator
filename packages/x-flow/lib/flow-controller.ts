import { XFlowUnit } from './flow-unit';

type FuncType = (...args: unknown[]) => unknown;

function compose(funcs: FuncType[]): FuncType {
  return funcs.reduce((accum, current) => {
    return async function (data: unknown) {
      const nextData: any = await accum(data);
      return current(nextData);
    };
  });
}

export class XFlowController {
  private _units: XFlowUnit[];

  constructor() {
    this._units = [];
  }

  addUnit(unit: XFlowUnit): void {
    this._units.push(unit);
  }

  async run(data?: any): Promise<any> {
    const final = compose(this._units.map(unit => unit.doWork.bind(unit)));

    const result = final(data);
    return result;
  }
}
