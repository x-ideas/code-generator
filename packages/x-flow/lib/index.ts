export * from './concreate-units/generate-service-request';
