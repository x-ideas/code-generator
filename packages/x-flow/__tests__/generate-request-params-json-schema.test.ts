import { OpenAPIV2 } from 'openapi-types';
import { CollectOutputFlowUnit } from '../lib/collect-output-flow-unit';
import { GenerateRequestParamsJsonSchemaFlowUnit } from '../lib/concreate-units/generate-request-params-json-schema';
import { ParseRequestCodeFlowUnit } from '../lib/concreate-units/parse-request-code';
import { SpecialOutputFlowUnit } from '../lib/concreate-units/special-output-flow-unit';
import { SwaggerParseFlowUnit } from '../lib/concreate-units/swagger-parse';
import { XFlowController } from '../lib/flow-controller';

import OpenAPIData from './assets/openAPI.json';

describe('测试GenerateRequestParamsJsonSchemaFlowUnit', () => {
  test('解析3610401', async () => {
    const fc = new XFlowController();

    const collectOutputUnit = new CollectOutputFlowUnit();

    const outputCodeUnit = new SpecialOutputFlowUnit('3610401');
    const outputOpenAPIUnit = new SpecialOutputFlowUnit(OpenAPIData);
    collectOutputUnit.addUnit(outputCodeUnit);
    collectOutputUnit.addUnit(outputOpenAPIUnit);

    const codeParseUnit = new ParseRequestCodeFlowUnit();
    const swaggerParseUnit = new SwaggerParseFlowUnit();

    const frpUnit = new GenerateRequestParamsJsonSchemaFlowUnit();

    fc.addUnit(collectOutputUnit);
    fc.addUnit(codeParseUnit);
    fc.addUnit(swaggerParseUnit);
    fc.addUnit(frpUnit);

    const result: OpenAPIV2.Document = await fc.run();
    // console.log(result);
  });
});
