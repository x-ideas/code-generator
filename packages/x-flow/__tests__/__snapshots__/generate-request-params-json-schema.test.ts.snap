// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试GenerateRequestParamsJsonSchemaFlowUnit path_body3610404 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "确认需求单，变更补货数: 3610404",
  "properties": Object {
    "id": Object {
      "description": "id",
      "type": "string",
    },
  },
  "required": Array [
    "id",
  ],
}
`;

exports[`测试GenerateRequestParamsJsonSchemaFlowUnit path_body3610404 2`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "确认需求单，变更补货数: 3610404",
  "properties": Object {},
  "required": Array [],
}
`;

exports[`测试GenerateRequestParamsJsonSchemaFlowUnit path_body3610404 3`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "确认需求单，变更补货数: 3610404",
  "properties": Object {
    "list": Object {
      "description": "list",
      "items": Object {
        "description": "补货需求单补货数",
        "properties": Object {
          "product_id": Object {
            "description": "商品id",
            "type": "string",
          },
          "replenish_count": Object {
            "description": "补货数",
            "type": "integer",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": Array [],
}
`;

exports[`测试GenerateRequestParamsJsonSchemaFlowUnit 不包含body和path 1`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "分页查找补货需求单: 3610401",
  "properties": Object {},
  "required": Array [],
}
`;

exports[`测试GenerateRequestParamsJsonSchemaFlowUnit 不包含body和path 2`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "分页查找补货需求单: 3610401",
  "properties": Object {},
  "required": Array [],
}
`;

exports[`测试GenerateRequestParamsJsonSchemaFlowUnit 不包含body和path 3`] = `
Object {
  "$schema": "http://json-schema.org/draft-04/schema#",
  "description": "分页查找补货需求单: 3610401",
  "properties": Object {
    "applyNo": Object {
      "description": "需求单号",
      "type": "string",
    },
    "page": Object {
      "description": "第几页(0..N)",
      "type": "integer",
    },
    "size": Object {
      "description": "每页大小",
      "type": "integer",
    },
    "sort": Object {
      "description": "排序，格式: property(,asc|desc). 默认升序. ",
      "items": Object {
        "description": "item-暂无",
        "type": "string",
      },
      "type": "array",
    },
    "status": Object {
      "description": "状态,-1=全部，1=草稿，10=待发送，20=待发货，30=待收货，40=到货中，50=已完成 ，120=未完成  200=删除",
      "type": "integer",
    },
    "storeId": Object {
      "description": "门店id",
      "type": "string",
    },
    "timeCreateFrom": Object {
      "description": "创建时间开始",
      "type": "integer",
    },
    "timeCreateTo": Object {
      "description": "创建时间结束",
      "type": "integer",
    },
  },
  "required": Array [
    "page",
    "size",
  ],
}
`;
